Description: Isolate sudo working environment before fixing timezone
Origin: backport, https://github.com/sudo-project/sudo/commit/1820a349687522f51023d1ae5925125f59679a8c
Bug: https://security-tracker.debian.org/tracker/CVE-2026-96512
Forwarded: not-needed

--- a/src/sudo.c
+++ b/src/sudo.c
@@ -163,6 +163,8 @@
     bindtextdomain(PACKAGE_NAME, LOCALEDIR);
     textdomain(PACKAGE_NAME);
 
+    /* Use the system timezone for sudoers time restrictions. */
+    unsetenv("TZ");
     (void) tzset();
 
     /* Must be done before we do any password lookups */
@@ -344,10 +346,27 @@
 int
 os_init_common(int argc, char *argv[], char *envp[])
 {
+    extern char **environ;
+    int envc;
 #ifdef STATIC_SUDOERS_PLUGIN
     preload_static_symbols();
 #endif
     gc_init();
+
+    /* Keep the original envp available to plugins while changing environ. */
+    for (envc = 0; envp[envc] != NULL; envc++)
+        continue;
+    if (envc != 0) {
+        char **new_env = reallocarray(NULL, sizeof(char *), envc + 1);
+        if (new_env == NULL || !gc_add(GC_PTR, new_env)) {
+            sudo_fatalx_nodebug(U_("%s: %s"), __func__,
+                U_("unable to allocate memory"));
+        }
+        for (envc = 0; envp[envc] != NULL; envc++)
+            new_env[envc] = envp[envc];
+        new_env[envc] = NULL;
+        environ = new_env;
+    }
     return 0;
 }
 
